statistics/handle/statscache.go

```diff
@@ -15,15 +15,274 @@ package handle
+	GetAll() []*statistics.Table
+	Clear()
+	Close()
+	GetBytesLimit() int64
+	SetBytesLimit(bytesLimit int64)
+	BytesConsumed() int64
+}
+
+const byteShards = 8
+const numShards = 1 << byteShards
+
+// statsCache caches Regions loaded from PD.
+type ristrettoStatsCache struct {
+	mu         sync.Mutex
+	cache      *cache.Cache
+	version    uint64
+	memTracker *memory.Tracker // track memory usage.
+	tablesMap
+}
+
+type tablesMap struct {
+	tablesShards [numShards]tableShard
+}
+
+// tableShard is a shard of table
+type tableShard struct {
+	data map[int64]*statistics.Table
+	sync.RWMutex
+}
+
+// Set set key with value
+func (ts *tablesMap) Set(key int64, value *statistics.Table) {
+	shard := &ts.tablesShards[key&(numShards-1)]
+	shard.Lock()
+	defer shard.Unlock()
+	shard.data[key>>byteShards] = value
+}
+
+// Get get key with value table
+func (ts *tablesMap) Get(key int64) (*statistics.Table, bool) {
+	shard := &ts.tablesShards[key&(numShards-1)]
+	shard.RLock()
+	defer shard.RUnlock()
+	data, ok := shard.data[key>>byteShards]
+	return data, ok
+}
+
+// Del delete key
+func (ts *tablesMap) Del(key int64) {
+	shard := &ts.tablesShards[key&(numShards-1)]
+	shard.Lock()
+	defer shard.Unlock()
+	delete(shard.data, key>>byteShards)
+}
+
+type statsCacheType int8
+
+const (
+	//RistrettoStatsCacheType type
+	RistrettoStatsCacheType statsCacheType = iota
+	//SimpleStatsCacheType simple type
+	SimpleStatsCacheType
+)
+
+// DefStatsCacheType defines statistic cache type
+var DefStatsCacheType = RistrettoStatsCacheType
+
+// NewStatsCache returns a new statsCahce with capacity memoryLimit(initial 1G)
+func NewStatsCache(memoryLimit int64, tp statsCacheType) (StatsCache, error) {
+	switch tp {
+	case RistrettoStatsCacheType:
+		return newRistrettoStatsCache(memoryLimit)
+	case SimpleStatsCacheType:
+		return newSimpleStatsCache(memoryLimit), nil
+	}
+	return nil, errors.New("wrong statsCache type")
+}
+
+// newRistrettoStatsCache returns a new statsCahce with capacity memoryLimit(initial 1G)
+func newRistrettoStatsCache(memoryLimit int64) (*ristrettoStatsCache, error) {
+	// since newRistrettoStatsCache controls the memory usage by itself, set the capacity of
+	// the underlying LRUCache to max to close its memory control
+	sc := &ristrettoStatsCache{
+		memTracker: memory.NewTracker(memory.LabelForStatsCache, -1),
+	}
+	for i := range sc.tablesShards {
+		sc.tablesShards[i].data = make(map[int64]*statistics.Table)
+	}
+	sc.version = 0
+	cache, err := cache.NewCache(&cache.Config{
+		NumCounters: 1e7,         // number of keys to track frequency of (10M).
+		MaxCost:     memoryLimit, // maximum cost of cache (1GB).
+		BufferItems: 64,          // number of keys per Get buffer.
+		Cost: func(value interface{}) int64 {
+			return (value.(*statistics.Table).MemoryUsage())
+		},
+		OnEvict: func(key uint64, value interface{}) {
+			if t, ok := sc.Get(int64(key)); ok {
+				if value != nil && t.PhysicalID == int64(key) {
```

Comment:
When the condition  `t.PhysicalID == int64(key)` will be false?

Reply:
i'll remove this.